File: arrays/removeDuplicatesFromSortedArray.java
package arrays;

public class removeDuplicatesFromSortedArray {
    public static void main(String[] args) {
        SolutionRDFSA solution = new SolutionRDFSA();
        System.out.println(solution.removeDuplicatesApproach2(new int[]{0,0,1,1,1,2,2,3,3,4}));
        System.out.println(solution.removeDuplicatesApproach2(new int[]{1,2,3,4}));
        System.out.println(solution.removeDuplicatesApproach2(new int[]{1,1,2}));
        System.out.println(solution.removeDuplicatesApproach2(new int[]{1}));
        System.out.println(solution.removeDuplicatesApproach2(new int[]{}));


    }
}

class SolutionRDFSA {
    public int removeDuplicates(int[] nums) {
        
        int i =0;
        int j =0;
        
        while(j< nums.length){
            while(j< nums.length && nums[i]==nums[j]) j++;
            i++;
            if(j< nums.length) nums[i]=nums[j];

        }
        
        return i;
    }


    // TC: O(N)
    // SC: O(1)
    public int removeDuplicatesApproach2(int[] nums) {
        
        int i =0;
        for (int j = 0 ; j< nums.length ; j++){
            if(i<nums.length && j<nums.length &&nums[i]!=nums[j]){
                nums[++i]=nums[j];
            }
        }
        return i+1;
    }
}

File: arrays/removeDuplicatesFromSortedArray.py
# https://leetcode.com/problems/remove-duplicates-from-sorted-array/

# Runtime: 138 ms, faster than 47.61% of Python3 online submissions for Remove Duplicates from Sorted Array.
# Memory Usage: 15.6 MB, less than 21.60% of Python3 online submissions for Remove Duplicates from Sorted Array.


class Solution:
    def removeDuplicates(self, nums: list[int]) -> int:
        k = 0
        for i in range(len(nums)):
            if i < len(nums) - 1 and nums[i] == nums[i + 1]:
                continue
            nums[k] = nums[i]
            k += 1
        return k
